### PR TITLE
At 5585 convert serverless log forwarder to typescript

### PR DESCRIPTION
**Pull request by htorianik**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,32 +1,45 @@
 {
   "name": "serverless-log-forwarding",
   "version": "2.0.0",
```
Comment (rddimon) [inline package.json]:
```suggestion
  "version": "3.0.0",
```

File: package.json [hunk 2]
```diff
@@ -1,32 +1,45 @@
...
+  },
   "description": "a serverless plugin to forward logs to given lambda function",
-  "main": "index.js",
+  "main": "dist/src/index.js",
   "directories": {
     "test": "test"
   },
   "dependencies": {
-    "underscore": "^1.8.3"
+    "@types/underscore": "^1.11.4",
```
Comment (rddimon) [inline package.json:13]: For what is needed `@types/underscore`?

File: package.json [hunk 3]
```diff
@@ -1,32 +1,45 @@
...
-    "shelljs": "^0.8.5"
+    "serverless": "<4",
+    "shelljs": "^0.8.5",
+    "ts-node": "^10.4.0",
+    "typescript": "^4.5.5",
+    "eslint": "^8.7.0",
+    "@typescript-eslint/parser": "^5.11.0",
+    "@typescript-eslint/eslint-plugin": "^5.11.0",
+    "eslint-config-airbnb": "^19.0.4",
+    "eslint-config-airbnb-typescript": "^16.1.0"
```
Comment (rddimon) [inline package.json]: You can do alphabetical ordering of `devDependencies` it will be better to read

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  FunctionName: string;
+}
+
+interface SubscriptionFilterProps {
+  DestinationArn: string;
+  FilterPattern: string;
+  LogGroupName: string;
+  RoleArn?: string;
+}
+
```
Comment (rddimon) [inline src/index.ts:37]: I would suggest moving interfaces to the `types.ts`.
Those are abstractions and can be used in different places.

File: package.json [hunk 5]
```diff
@@ -1,32 +1,48 @@
 {
   "name": "serverless-log-forwarding",
-  "version": "2.0.0",
+  "version": "3.0.0",
+  "engines": {
+    "node": ">=14"
```
Comment (rddimon) [inline package.json:5]:
```suggestion
    "node": ">=12"
```

https://nodejs.org/en/about/releases/
The end of life is 2022-04-30 for node 12 so it's better to support for now

File: package.json (hunk 2 above)
Comment (rddimon) [reply package.json:13]: As I understand the prefix `@types` it's something related to the typescript code validation/highlighting, right?

File: src/index.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+interface PluginConfig {
+  stages?: string[];
+  roleArn?: string;
+  filterPattern: string;
+  normalizedFilterID: boolean;
+  createLambdaPermission: boolean;
+  destinationARN: string;
+}
+
+const configDefaults = {
```
Comment (rddimon) [inline src/index.ts:13]:
```suggestion
const CONFIG_DEFAULTS = {
```
Or you can make it as a part of the plugin class

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+  DestinationArn: string;
+  FilterPattern: string;
+  LogGroupName: string;
+  RoleArn?: string;
+}
+
+type LambdaPermission = CFObject<LambdaPermissionProps>;
+type SubscriptionFilter = CFObject<SubscriptionFilterProps>;
+
+function getStage(options: ServerlessConfig, sls: ServerlessInstance) {
```
Comment (rddimon) [inline src/index.ts:41]: This function I would suggest making as a part of the class or create a separate file like `utils.ts` to declare such helper functions

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    this.serverless = serverless;
+    this.options = options;
+    this.provider = this.serverless.getProvider('aws');
+    this.hooks = {
+      'package:initialize': this.updateResources.bind(this),
+    };
+  }
+
+  loadConfig() {
+    this.config = { ...configDefaults, ...this.serverless.service.custom.logForwarding };
```
Comment (rddimon) [inline src/index.ts:69]: You need to add a validation here for example check `this.serverless.service.custom` option is specified

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    }
+    return service.resources.Resources;
+  }
+
+  createResourcesObj(): object {
+    const { service } = this.serverless;
+    const resourceObj = {};
+
+    const createLambdaPermission = this.config.createLambdaPermission && !this.config.roleArn;
+    const permissionId = 'LogForwardingLambdaPermission';
```
Comment (rddimon) [inline src/index.ts:110]: Better to make `LogForwardingLambdaPermission` global variable

File: package.json (hunk 2 above)
Comment (htorianik) [reply package.json:13]: addressed

File: package.json (hunk 3 above)
Comment (htorianik) [reply package.json]: addressed

File: src/index.ts (hunk 4 above)
Comment (htorianik) [reply src/index.ts:37]: addressed

File: CHANGELOG.md [hunk 10]
```diff
@@ -7,6 +7,13 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+## [3.0.0] - 2022-02-03
+
+### Added
+
+- Moved the plugin and tests to TypeScript.
+- Moved from istanbul to nyc.
```
Comment (aoskotsky-amplify) [inline CHANGELOG.md:15]: 🤣

File: package.json (hunk 5 above)
Comment (htorianik) [reply package.json:5]: addressed

File: src/index.ts (hunk 6 above)
Comment (htorianik) [reply src/index.ts:13]: addressed

File: src/index.ts (hunk 7 above)
Comment (htorianik) [reply src/index.ts:41]: addressed

File: src/index.ts (hunk 8 above)
Comment (htorianik) [reply src/index.ts:69]: addressed

File: src/index.ts (hunk 9 above)
Comment (htorianik) [reply src/index.ts:110]: addressed

File: src/index.ts (hunk 8 above)
Comment (htorianik) [reply src/index.ts:69]: addressed

File: src/index.ts (hunk 9 above)
Comment (htorianik) [reply src/index.ts:110]: addressed

**Comment by rddimon** [comment]
`test/integration-tests/fixtures/resources/logs-receiver.js` should this file be with `.js` extension?

File: src/index.ts [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    const { stage } = this.options;
+    if (stage && stage !== '') {
+      return stage;
+    }
+    return this.serverless.service.provider.stage;
+  }
+
+  makeSubsctiptionFilter(functionName: string, deps?: string[]): SubscriptionFilterCF {
+    const functionObject = this.serverless.service.getFunction(functionName);
+    const logGroupName = this.provider.naming.getLogGroupName(functionObject.name);
```
[the remark this answers is not in the document]
Comment (htorianik) [reply src/index.ts:120]: Yes, it is. I don't know the details but so was done it the original code and it fails if I use `functionName`

File: src/index.ts [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    };
+  }
+
+  makeLambdaPermission(): LambdaPermissionCF {
+    const { region } = this.serverless.service.provider;
+    const principal = `logs.${region}.amazonaws.com`;
+    return {
+      Type: 'AWS::Lambda::Permission',
+      Properties: {
+        FunctionName: this.config.destinationARN,
```
[the remark this answers is not in the document]
Comment (htorianik) [reply src/index.ts:141]: It has already already been checked by `loadConfig`

File: package.json [hunk 13]
```diff
@@ -1,47 +1,61 @@
...
-    "serverless",
-    "serverless",
-    "plugin",
-    "logs"
-  ],
-  "author": "Amplify Education",
-  "license": "MIT",
-  "bugs": {
-    "url": "https://github.com/amplify-education/serverless-log-forwarding/issues"
+    "serverless": "<4",
```
Comment (rddimon) [inline package.json:53]:
```suggestion
    "serverless": "^3.1.0",
```
Also, you can add 
```
"peerDependencies": {
    "serverless": "<4"
  }
```
To support sls version 1 | 2 | 3

As we did for the Domain Manager https://github.com/amplify-education/serverless-domain-manager/blob/master/package.json

File: src/index.ts [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+    const createLambdaPermission = this.config.createLambdaPermission && !this.config.roleArn;
+    if (createLambdaPermission) {
+      const permission = this.makeLambdaPermission();
+      resourceObj[PERMISSION_ID] = permission;
+    }
+
+    _.keys(service.functions)
+      .filter((functionName) => {
+        const { logForwarding = {} } = this.serverless.service.getFunction(functionName);
+        return !(logForwarding.enabled === false);
```
[the remark this answers is not in the document]
Comment (htorianik) [reply src/index.ts:92]: It's not the same. If `logForwarding.enabled === undefined`, then the result should be `true`, in your example it's false

File: package.json [hunk 15]
```diff
@@ -1,47 +1,61 @@
...
+    "shelljs": "^0.8.5",
+    "ts-node": "^10.4.0",
+    "typescript": "^4.5.5"
   },
-  "homepage": "https://github.com/amplify-education/serverless-log-forwarding#readme"
+  "engines": {
+    "node": ">=12"
+  }
```
[the remark this answers is not in the document]
Comment (htorianik) [reply package.json:60]: I sorted package.json using `sort-package-json` package. Which is quite popular. I like it and it is standartized. I'd like to leave it as it is. Hope you don't mind.

File: src/index.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (rddimon) [reply src/index.ts:92]: `!!logForwarding.enabled` will work

**Review by rddimon — APPROVED** [review]
GJ. Looks fine!

**Pull request merged**